examples/custom_search_method/local_search_runner/asha.py

```diff
@@ -0,0 +1,480 @@
+are invoked by LocalSearchRunner in response to the SearcherEvents received
+from the multi-trial experiment. The methods return a list of Operations to LocalSearchRunner
+which sends them to the multi-trial experiment for execution, as depicted below.
+
+Multi-trial experiment  --- (SearcherEvent1) ----> LocalSearchRunner  --- (SearcherEvent1) ---> SearchMethod
+Multi-trial experiment <---   (Operations1)  ----  LocalSearchRunner <---   (Operations1)  ---  SearchMethod
+Multi-trial experiment  --- (SearcherEvent2) ----> LocalSearchRunner  --- (SearcherEvent2) ---> SearchMethod
+Multi-trial experiment <---   (Operations2)  ----  LocalSearchRunner <---   (Operations2)  ---  SearchMethod
+and so on.
+
+Currently, we support the following operations:
+  -> Create - starts a new trial with a unique trial id and a set of hyperparameters,
+  -> ValidateAfter - sets number of steps (i.e., batches or epochs) after which a validation is run for a trial
+                     with a given id,
+  -> Close - closes a trial with a given id,
+  -> Shutdown - closes the experiment.
+
+To support fault tolerance, a custom SearchMethod has to implement save_method_state() and load_method_state(), which
+provide logic for saving and loading any information internal to the SearchMethod,
+such as variables, structures, or models, that are modified as the SearchMethod progresses.
+save_method_state() and load_method_state() are called by LocalSearchRunner as a part of LocalSearchRunner.save() and
+LocalSearchRunner.load() to ensure that LocalSearchRunner state and SearchMethod state can be restored if
+LocalSearchRunner is terminated or interrupted.
+To provide fault tolerance, LocalSearchRunner calls save() and save_method_state() after processing
+each SearcherEvent and before sending the list of Operations to the multi-trial experiment.
+On resumption, LocalSearchRunner calls load() and load_method_state(), and gets SearcherEvents from the multi-trial
+experiment. If a SearcherEvent has already been processed, then LocalSearchRunner re-sends the operations.
+Otherwise, it continues to process SearcherEvents. That is, LocalSearchRunner invokes the SearchMethod methods
+corresponding to the Searcher Event, gets the operations, saves the state, and sends the operations.
+
+While implementation of save_method_state() and load_method_state() depends on the user,
+in this example we chose to encapsulate all variables required by ASHASearchMethod in a new class,
+ASHASearchMethodState. We elected to use pickle as the storage format for convenience.
+'''
+
+import dataclasses
+import logging
+import random
+import pickle
+import sys
+import uuid
+from pathlib import Path
+from typing import Dict, List, Set
+
+from determined.searcher.search_method import (
+    Close,
+    Create,
+    ExitedReason,
+    Operation,
+    SearchMethod,
+    Shutdown,
+    ValidateAfter,
+)
+
+
+class ASHASearchMethod(SearchMethod):
```

Comment:
is there a difference between `asha.py` for local runner and for core api runner?  if not, we may want to restructure so we have a single example with subdirectories for local and core api but shared asha.py file in the main level directory.